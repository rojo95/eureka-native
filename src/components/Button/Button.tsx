import React, { ReactNode } from "react";
import {
    Pressable,
    StyleSheet,
    Text,
    TouchableOpacityProps,
} from "react-native";
import {
    DefaultTheme,
    Button as PaperButton,
    useTheme,
} from "react-native-paper";
import { StyleProps } from "react-native-reanimated";

export const BUTTON_TYPES = {
    PRIMARY: 1,
    SECONDARY: 2,
    LINK: 3,
} as const;

interface ButtonPropsBase extends TouchableOpacityProps {
    type?: (typeof BUTTON_TYPES)[keyof typeof BUTTON_TYPES];
    icon?: ReactNode | undefined;
    text?: string;
    textStyle?: StyleProps;
    buttonStyle?: StyleProps;
    children?: ReactNode;
}

// Extends the base interface to require the field 'text'.
interface WithText extends ButtonPropsBase {
    icon?: ReactNode;
    text: string;
    children?: ReactNode;
}

// Extends the base interface to require the 'children' field
interface WithChildren extends ButtonPropsBase {
    icon?: ReactNode;
    text?: string;
    children: ReactNode;
}

// Extends the base interface to require the 'icon' field
interface WithIcon extends ButtonPropsBase {
    icon: ReactNode;
    text?: string;
    children?: ReactNode;
}

type ButtonProps = WithText | WithChildren | WithIcon;

export default function Button({
    type = BUTTON_TYPES.PRIMARY,
    onPress,
    icon,
    text,
    textStyle,
    buttonStyle,
    children,
    disabled,
}: ButtonProps) {
    const theme: DefaultTheme = useTheme();

    const styles = StyleSheet.create({
        button: {
            borderColor:
                type === BUTTON_TYPES.LINK
                    ? "transparent"
                    : disabled
                    ? theme.colors.primaryLight
                    : theme.colors.primary,
            borderWidth: 1,
            backgroundColor:
                type === BUTTON_TYPES.PRIMARY
                    ? disabled
                        ? theme.colors.primaryLight
                        : theme.colors.primary
                    : type === BUTTON_TYPES.LINK
                    ? "transparent"
                    : theme.colors.primaryContrast,
            borderRadius: 5,
            width: "100%",
        },
        text: {
            color:
                type === BUTTON_TYPES.PRIMARY
                    ? theme.colors.primaryContrast
                    : disabled
                    ? theme.colors.primaryLight
                    : theme.colors.primary,
            width: "100%",
        },
    });

    if (type === BUTTON_TYPES.LINK) {
        return (
            <Pressable
                disabled={disabled}
                onPress={onPress}
                style={[styles.button, buttonStyle]}
            >
                {icon && icon}
                {text && <Text style={[styles.text, textStyle]}>{text}</Text>}
                {children && children}
            </Pressable>
        );
    }

    return (
        <PaperButton
            style={[styles.button, buttonStyle]}
            icon={() => icon}
            onPress={onPress}
            disabled={disabled}
        >
            {text && <Text style={[styles.text, textStyle]}>{text}</Text>}
            {children && children}
        </PaperButton>
    );
}
