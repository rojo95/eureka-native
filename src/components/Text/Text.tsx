import { customText } from "react-native-paper";

// Use this instead of importing `Text` from `react-native-paper`
const Text = customText<"customVariant">();

export default Text;
