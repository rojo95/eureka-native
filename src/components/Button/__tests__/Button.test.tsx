import React from "react";
import renderer from "react-test-renderer";

import Button, { BUTTON_TYPES } from "../Button";
import { Button as ButtonRNP } from "react-native-paper";
import Text from "components/Text/Text";

describe("Test to evaluate <Button /> component", () => {
    it("has 1 child", () => {
        const tree: any = renderer.create(<Button text="text" />).toJSON();
        expect(tree.children.length).toBe(1);
    });

    it("render correctly each kind of button", () => {
        const link: any = renderer
            .create(<Button text="text" type={BUTTON_TYPES.LINK} />)
            .toJSON();
        expect(link.children.length).toBe(1);

        const primary: any = renderer
            .create(<Button text="text" type={BUTTON_TYPES.PRIMARY} />)
            .toJSON();
        expect(primary.children.length).toBe(1);

        const secondary: any = renderer
            .create(<Button text="text" type={BUTTON_TYPES.SECONDARY} />)
            .toJSON();
        expect(secondary.children.length).toBe(1);
    });

    it("renders correctly with text", () => {
        const testText = "test";
        const tree: any = renderer.create(<Button text={testText} />).toJSON();
        const textElement =
            tree.children[0]?.children[0]?.children[0]?.children[1]
                ?.children[0];
        expect(textElement.children).toContain(testText);
    });

    it("renders correctly with icon", () => {
        const textInsideButtonIcon = "Press Me";
        const tree: any = renderer
            .create(
                <Button
                    icon={
                        <ButtonRNP icon="camera">
                            {textInsideButtonIcon}
                        </ButtonRNP>
                    }
                />
            )
            .toJSON();

        const icon =
            tree?.children[0]?.children[0]?.children[0]?.children[0]
                ?.children[0]?.children[0]?.children[0]?.children[0]?.children;

        expect(icon[0].props.testID).toBe("button-icon-container");
        expect(icon[1]?.children[0]).toBe(textInsideButtonIcon);
    });

    it("should render children correctly", () => {
        const childrenInsideButton = "Text Button test";
        const tree: any = renderer
            .create(
                <Button>
                    <Text>{childrenInsideButton}</Text>
                </Button>
            )
            .toJSON();
        const text =
            tree?.children[0]?.children[0]?.children[0]?.children[1]
                ?.children[0];
        expect(text.children).toContain(childrenInsideButton);
    });

    it("render button without change", () => {
        const tree = renderer.create(<Button text="text" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
