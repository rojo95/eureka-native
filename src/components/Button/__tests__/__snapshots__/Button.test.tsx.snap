// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to evaluate <Button /> component render button without change 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(103, 80, 164, 1)",
      "borderRadius": 5,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 0,
      "width": "100%",
    }
  }
  testID="button-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(103, 80, 164, 1)",
        "borderColor": "rgba(103, 80, 164, 1)",
        "borderRadius": 5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flex": undefined,
        "minWidth": 64,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="button-container"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 5,
          },
        ]
      }
      testID="button"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "marginLeft": 12,
                "marginRight": -4,
              },
              {
                "marginLeft": 16,
                "marginRight": -16,
              },
              {
                "marginLeft": 12,
                "marginRight": -8,
              },
            ]
          }
          testID="button-icon-container"
        />
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 0.1,
                  "lineHeight": 20,
                },
                [
                  {
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  false,
                  {
                    "marginHorizontal": 16,
                  },
                  undefined,
                  false,
                  {
                    "color": "rgba(103, 80, 164, 1)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  undefined,
                ],
              ],
            ]
          }
          testID="button-text"
        >
          <Text
            style={
              [
                {
                  "color": undefined,
                  "width": "100%",
                },
                undefined,
              ]
            }
          >
            text
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
